### PR TITLE
WIP: Kotlinjs compiler

### PR DESCRIPTION
**Pull request by Foso**
#22 

I'm able to run the KotlinJSCompiler with my plugin. I need to add a useful default configuration for K2JSCompilerArguments and remove unused code

### CODE REVIEW

File: src/main/kotlin/com/tschuchort/compiletesting/KaptComponentRegistrar.kt [hunk 1]
```diff
@@ -40,32 +40,30 @@ import org.jetbrains.kotlin.kapt3.AbstractKapt3Extension
...
 import org.jetbrains.kotlin.resolve.jvm.extensions.AnalysisHandlerExtension
 import java.io.File
-import javax.annotation.processing.Processor
-
-internal class KaptComponentRegistrar : ComponentRegistrar {
 
+class KaptComponentRegistrar: ComponentRegistrar {
     override fun registerProjectComponents(project: MockProject, configuration: CompilerConfiguration) {
-        if (threadLocalParameters.get().processors.isEmpty())
+        if (CommonComponentRegistrar.threadLocalParameters.get().processors.isEmpty())
```
Comment (tschuchortdev) [inline src/main/kotlin/com/tschuchort/compiletesting/KaptComponentRegistrar.kt]: Why not just pass those as constructor parameters, now that we instantiate the `KaptComponentRegistrar` manually?

File: src/main/kotlin/com/tschuchort/compiletesting/KotlinCompilation.kt [hunk 2]
```diff
@@ -54,7 +59,7 @@ class KotlinCompilation {
 		log("Created temporary working directory at ${path.toAbsolutePath()}")
 		return@default path.toFile()
 	}
-
+	var componentRegistrars: List<ComponentRegistrar> = emptyList()
```
Comment (tschuchortdev) [inline src/main/kotlin/com/tschuchort/compiletesting/KotlinCompilation.kt]: I would prefer the name `compilerPlugins`.

File: src/main/kotlin/com/tschuchort/compiletesting/KotlinCompilation.kt [hunk 3]
```diff
@@ -502,19 +493,19 @@ class KotlinCompilation {
...
-	private fun compileKotlin(sourceFiles: List<File>): ExitCode {
+	private fun compileKotlin(sourceFiles: List<File>): com.tschuchort.compiletesting.ExitCode {
 		val sources = sourceFiles +
 				kaptKotlinGeneratedDir.listFilesRecursively() +
 				kaptSourceDir.listFilesRecursively()
 
 		// if no Kotlin sources are available, skip the compileKotlin step
 		if(sources.filter<File>(File::hasKotlinFileExtension).isEmpty())
-			return ExitCode.OK
+			return com.tschuchort.compiletesting.ExitCode.OK
```
Comment (tschuchortdev) [inline src/main/kotlin/com/tschuchort/compiletesting/KotlinCompilation.kt]: Can be imported directly to remove the needless package names everywhere.

**Comment by tschuchortdev** [comment]
Very nice. Can we integrate the `K2JSCompiler` with the regular `KotlinCompilation` so it's possible to compile Kotlin, Java and JavaScript at once?

**Comment by eiswind** [comment]
Are there any plans to integrate this PR? It seems to be stalled.
It could be really helpful in with some of my current "investigations". I did try to see it through, but was a little overwhelmed at first sight.

**Comment by tschuchortdev** [comment]
@eiswind Personally it's not on my agenda currently. I have a lot of other stuff on my plate right now and the upgrade to Kotlin 1.4 has a higher priority. But if you are willing to take a shot at it, I can assist you. Looking at the PR there is a lot of changes here that have already been integrated by now (the CommonComponentRegistrar for example) or are simple refactorings that you can ignore. The core of the changes necessary to support KotlinJS is in KotlinJSCompilation.kt which is largely copy&paste proof-of-concept code as well. If you diff that file with KotlinCompilation.kt you will find that there are only a few changes and it will not look as daunting anymore. 

I suspect that preliminary support of KotlinJS is pretty straight forward (just replace `K2JVMCompiler` with `K2JSCompiler` basically). The hard part will be ironing out any quirks that happen in different configurations and integrating it in the existing code in a modular way (if at all possible). The past few months have shown that I will have to rewrite the interface to make it more modular sooner or later.

**Comment by Foso** [comment]
Hi, unfortunately i couldn't find the time to keep on working on this. In the next days i will try see if my initial approach is still working with Kotlin 1.4.0 and otherwise change it, but i can't promise. If someone is interested in implementing this, i'll happy to help where i can.

**Comment by Foso** [comment]
HI, i just discovered that bnorm did contribute the KotlinJs Testing :) . Thanks to this, my draft is obsolete and i will close it

**Pull request closed without merging**